### PR TITLE
Programmer Docs

### PR DESCRIPTION
**Pull request by dgasmith**
Overhauls @andysim's programmer docs PR. As discussed in the previous PR, these docs could use a few small tweaks, but overall they move the documentation state forward. I would vote for getting these in and tweaking over time later.

Closes #1274.

## Status
- [x] Ready for review
- [x] Ready for merge

### CODE REVIEW

File: doc/sphinxman/source/prog_tour.rst [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+.. _`sec:prog_tour-exposing`:
+
+Exposing C++ code to Python
+---------------------------
+
+The recent push to move sections of the code that are not a bottleneck into the
+Python layer requires that the C++ code is callable from Python and that its
+results are accessible.  The result accessibility is addressed by populating
+the appropriate variables in the Wavefunction object.  To make the code
+callable from Python, we rely on the excellent :pybind11:`PyBind11 <>` library
```
Comment (loriab) [inline doc/sphinxman/source/prog_tour.rst:90]:
```suggestion
callable from Python, we rely on the excellent `PyBind11 <https://pybind11.readthedocs.io/en/stable/>`_ library
```
unless you know better

File: doc/sphinxman/source/prog_ways_to_add.rst [hunk 2]
```diff
@@ -35,15 +35,106 @@ Ways to Add Code: Psi4NumPy, Plugins, Full Integration
...
-integrated fully into Psi4, or stay as a Plugin.
+Since those early days, the code has undergone some important structural
+changes that have greatly simplified the development workflow.  These changes
+were motivated by the realization that only a few bottlenecks exist in a typical
+calculation; by focusing on optimized C++ implementations of these bottlenecks
+and making these C++ functions available in Python, most of the code to implement
+the overall calculation can be written in simpler Python code.  Python is far
+better suited to management tasks such as directory navigation and retrieval,
+making it a natural choice for overall calculation layout than C++.  With the
+emergence of :numpy:`NumPy <>` as a standard tool for executing almost any
```
Comment (loriab) [inline doc/sphinxman/source/prog_ways_to_add.rst:57]:
```suggestion
emergence of `NumPy <https://numpy.org/>`_ as a standard tool for executing almost any
```

File: doc/sphinxman/source/prog_ways_to_add.rst [hunk 3]
```diff
@@ -35,15 +35,106 @@ Ways to Add Code: Psi4NumPy, Plugins, Full Integration
...
+C++ to Python has facilitated a much simpler work flow for prototyping and
+developing methods: this is detailed in the next section.
 
-There is a separate section that more fully explains plugins.
+.. _`sec:prog_psi4numpy`:
 
+Rapid initial development using Psi4NumPy
+-----------------------------------------
 
+The :p4numpy:`Psi4NumPy <>` project [Smith:2018:3504]_ is the recommended
```
Comment (loriab) [inline doc/sphinxman/source/prog_ways_to_add.rst:67]:
```suggestion
The `Psi4NumPy <https://github.com/psi4/psi4numpy>`_ project [Smith:2018:3504]_ is the recommended
```

File: doc/sphinxman/source/prog_ways_to_add.rst [hunk 4]
```diff
@@ -35,15 +35,106 @@ Ways to Add Code: Psi4NumPy, Plugins, Full Integration
...
+algebra, Fourier transforms, and general tensor manipulations, a massive number
+of methods can be easily implemented very easily using that library.  To
+facilitate this workflow, |PSIfour| exports key quantities such as integrals,
+densities and molecular orbitals in NumPy format.  From this point, the
+programmer can simply call the appropriate |PSIfour| functions to compute the
+desired input quantities, retrieve them in NumPy format, and then write the
+remaining code using standard Python and/or NumPy syntax.  This approach does
+not require any recompilation of code, resulting in a particularly facile
+development workflow.  Detailed examples and tutorials are available in the
+:p4numpy:`Psi4NumPy <>` repository.
```
Comment (loriab) [inline doc/sphinxman/source/prog_ways_to_add.rst:79]:
```suggestion
`Psi4NumPy <https://github.com/psi4/psi4numpy>`_ repository.
```

File: doc/sphinxman/source/prog_ways_to_add.rst [hunk 5]
```diff
@@ -35,15 +35,106 @@ Ways to Add Code: Psi4NumPy, Plugins, Full Integration
...
 
-There is a separate section that more fully explains plugins.
+.. _`sec:prog_psi4numpy`:
 
+Rapid initial development using Psi4NumPy
+-----------------------------------------
 
+The :p4numpy:`Psi4NumPy <>` project [Smith:2018:3504]_ is the recommended
+mechanism for developing and prototyping new methods in Psi4.  Because
+:numpy:`NumPy <>` provides such a rich set of features for efficient linear
```
Comment (loriab) [inline doc/sphinxman/source/prog_ways_to_add.rst:69]:
```suggestion
`NumPy <https://numpy.org/>`_ provides such a rich set of features for efficient linear
```

**Review by loriab — APPROVED** [review]
hooray, programmer docs!

**Review by hokru — APPROVED** [review]

**Review by robertodr — APPROVED** [review]

**Pull request merged**